### PR TITLE
[LLVMCPU] Add pass to enable Armv9 Streaming SVE mode

### PR DESCRIPTION
**Pull request by c-rhodes**
This patch adds a pass 'iree-llvmcpu-enable-aarch64-ssve' that enables the Armv9 Scalable Matrix Extension (SME) Streaming SVE (SSVE) mode [1].

SSVE is enabled in the LLVM backend at the function boundary by specifying one of the following attributes [2]:

  aarch64_pstate_sm_enabled - calls to functions with this attribute are
                              wrapped with 'smstart sm' / 'smstop sm' [3].
                              Changes the function ABI.

  aarch64_pstate_sm_body    - 'smstart sm' / 'smstop sm' are emitted in the
                              function prologue / epilogue for functions
                              marked with this attribute. This is internal
                              and doesn't change the function ABI.

This pass adds the 'aarch64_pstate_sm_body' attribute to functions via the passthrough mechanism [4].

This attribute is used because PSTATE.SM changes are kept internal to the function and the purpose of this pass is to enable SSVE for dispatch functions which are called by the IREE runtime. The AAPCS64 [5] states it is the caller's responsibility to ensure that PSTATE.SM has a valid value on entry to a callee, the 'aarch64_pstate_sm_enabled' attribute would change the function ABI forcing the caller (IREE runtime) to be responsible for managing PSTATE.SM before entry/exit. At present, the runtime doesn't know the details of dispatches such that it could emit these instructions.

The pass is enabled for AArch64 when SVE(2) and SME are enabled for the following lowering configurations:

  * CPUBufferOpsTileAndVectorize
  * CPUDoubleTilingPeelingExpert
  * CPUConvTileAndDecomposeExpert

These configurations were chosen simply because they're used in one of our pipelines.

[1] https://developer.arm.com/documentation/ddi0616/aa
[2] https://llvm.org/docs/AArch64SME.html
[3] https://developer.arm.com/documentation/ddi0602/2023-03/Base-Instructions/SMSTART--Enables-access-to-Streaming-SVE-mode-and-SME-architectural-state--an-alias-of-MSR--immediate--
[4] https://mlir.llvm.org/docs/Dialects/LLVM/#attribute-pass-through
[5] https://github.com/ARM-software/abi-aa/blob/main/aapcs64/aapcs64.rst#671pstatesm-interfaces

### CODE REVIEW

**Comment by c-rhodes** [comment]
This is related to the RFC on Supporting Armv9 Scalable Matrix Extension (SME) Streaming SVE (SSVE) mode in MLIR / IREE, see #13556

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Thanks for this change. I think we shouldnt be adding this attribute to `func.func`. It should go into the `cpu_features` here https://github.com/openxla/iree/blob/a8e873be8a50e96d59a9a484e1c5479e2791aff0/compiler/src/iree/compiler/Codegen/LLVMCPU/test/pipeline_tests.mlir#L12 . I think then we dont need a special flag... but maybe we can add this to `iree-llvmcpu-target-cpu-features` (https://github.com/openxla/iree/blob/a8e873be8a50e96d59a9a484e1c5479e2791aff0/compiler/src/iree/compiler/Dialect/HAL/Target/LLVMCPU/LLVMTargetOptions.cpp#L91) . Example here https://github.com/openxla/iree/blame/a8e873be8a50e96d59a9a484e1c5479e2791aff0/tests/e2e/matmul/CMakeLists.txt#L63

**Comment by c-rhodes** [comment]
> Thanks for this change. I think we shouldnt be adding this attribute to `func.func`. It should go into the `cpu_features` here
> 
> https://github.com/openxla/iree/blob/a8e873be8a50e96d59a9a484e1c5479e2791aff0/compiler/src/iree/compiler/Codegen/LLVMCPU/test/pipeline_tests.mlir#L12
> . I think then we dont need a special flag... but maybe we can add this to `iree-llvmcpu-target-cpu-features` (
> 
> https://github.com/openxla/iree/blob/a8e873be8a50e96d59a9a484e1c5479e2791aff0/compiler/src/iree/compiler/Dialect/HAL/Target/LLVMCPU/LLVMTargetOptions.cpp#L91
> ) . Example here https://github.com/openxla/iree/blame/a8e873be8a50e96d59a9a484e1c5479e2791aff0/tests/e2e/matmul/CMakeLists.txt#L63

Thanks Mahesh. If the attribute isn't added to the function streaming mode wont be enabled. I suppose the pass that adds the attribute could be enabled if `--iree-llvmcpu-target-cpu-features="+sve,+sme"` is specified, rather than both this and `--iree-llvmcpu-enable-aarch64-ssve` when using iree-compile.

**Comment by MaheshRavishankar** [comment]
> > Thanks for this change. I think we shouldnt be adding this attribute to `func.func`. It should go into the `cpu_features` here
> > https://github.com/openxla/iree/blob/a8e873be8a50e96d59a9a484e1c5479e2791aff0/compiler/src/iree/compiler/Codegen/LLVMCPU/test/pipeline_tests.mlir#L12
> > 
> > . I think then we dont need a special flag... but maybe we can add this to `iree-llvmcpu-target-cpu-features` (
> > https://github.com/openxla/iree/blob/a8e873be8a50e96d59a9a484e1c5479e2791aff0/compiler/src/iree/compiler/Dialect/HAL/Target/LLVMCPU/LLVMTargetOptions.cpp#L91
> > 
> > ) . Example here [`a8e873b`/tests/e2e/matmul/CMakeLists.txt#L63 (blame)](https://github.com/openxla/iree/blame/a8e873be8a50e96d59a9a484e1c5479e2791aff0/tests/e2e/matmul/CMakeLists.txt#L63)
> 
> Thanks Mahesh. If the attribute isn't added to the function streaming mode wont be enabled. I suppose the pass that adds the attribute could be enabled if `--iree-llvmcpu-target-cpu-features="+sve,+sme"` is specified, rather than both this and `--iree-llvmcpu-enable-aarch64-ssve` when using iree-compile.

Instead of requiring it to be on the function, can it be on any parent op? When lowering to LLVM you could look for the attribute on any parent op and add that to the lowered LLVM function?
Irrespective of that, using target features to enable the pass would also work.

**Comment by c-rhodes** [comment]
> > > Thanks for this change. I think we shouldnt be adding this attribute to `func.func`. It should go into the `cpu_features` here
> > > https://github.com/openxla/iree/blob/a8e873be8a50e96d59a9a484e1c5479e2791aff0/compiler/src/iree/compiler/Codegen/LLVMCPU/test/pipeline_tests.mlir#L12
> > > 
> > > . I think then we dont need a special flag... but maybe we can add this to `iree-llvmcpu-target-cpu-features` (
> > > https://github.com/openxla/iree/blob/a8e873be8a50e96d59a9a484e1c5479e2791aff0/compiler/src/iree/compiler/Dialect/HAL/Target/LLVMCPU/LLVMTargetOptions.cpp#L91
> > > 
> > > ) . Example here [`a8e873b`/tests/e2e/matmul/CMakeLists.txt#L63 (blame)](https://github.com/openxla/iree/blame/a8e873be8a50e96d59a9a484e1c5479e2791aff0/tests/e2e/matmul/CMakeLists.txt#L63)
> > 
> > 
> > Thanks Mahesh. If the attribute isn't added to the function streaming mode wont be enabled. I suppose the pass that adds the attribute could be enabled if `--iree-llvmcpu-target-cpu-features="+sve,+sme"` is specified, rather than both this and `--iree-llvmcpu-enable-aarch64-ssve` when using iree-compile.
> 
> Instead of requiring it to be on the function, can it be on any parent op? When lowering to LLVM you could look for the attribute on any parent op and add that to the lowered LLVM function? Irrespective of that, using target features to enable the pass would also work.

Is there some advantage to it being on the parent op (typically a module I guess?) you have in mind?

**Comment by MaheshRavishankar** [comment]
> > > > Thanks for this change. I think we shouldnt be adding this attribute to `func.func`. It should go into the `cpu_features` here
> > > > https://github.com/openxla/iree/blob/a8e873be8a50e96d59a9a484e1c5479e2791aff0/compiler/src/iree/compiler/Codegen/LLVMCPU/test/pipeline_tests.mlir#L12
> > > > 
> > > > . I think then we dont need a special flag... but maybe we can add this to `iree-llvmcpu-target-cpu-features` (
> > > > https://github.com/openxla/iree/blob/a8e873be8a50e96d59a9a484e1c5479e2791aff0/compiler/src/iree/compiler/Dialect/HAL/Target/LLVMCPU/LLVMTargetOptions.cpp#L91
> > > > 
> > > > ) . Example here [`a8e873b`/tests/e2e/matmul/CMakeLists.txt#L63 (blame)](https://github.com/openxla/iree/blame/a8e873be8a50e96d59a9a484e1c5479e2791aff0/tests/e2e/matmul/CMakeLists.txt#L63)
> > > 
> > > 
> > > Thanks Mahesh. If the attribute isn't added to the function streaming mode wont be enabled. I suppose the pass that adds the attribute could be enabled if `--iree-llvmcpu-target-cpu-features="+sve,+sme"` is specified, rather than both this and `--iree-llvmcpu-enable-aarch64-ssve` when using iree-compile.
> > 
> > Instead of requiring it to be on the function, can it be on any parent op? When lowering to LLVM you could look for the attribute on any parent op and add that to the lowered LLVM function? Irrespective of that, using target features to enable the pass would also work.
> 
> Is there some advantage to it being on the parent op (typically a module I guess?) you have in mind?

Actually nevermind, I think I was mistaken. Controlling this through CPU features makes sense. I don't think we need another flag

**Comment by c-rhodes** [comment]
> I suppose the pass that adds the attribute could be enabled if `--iree-llvmcpu-target-cpu-features="+sve,+sme"` is specified, rather than both this and `--iree-llvmcpu-enable-aarch64-ssve` when using iree-compile.

I got confused myself here, this is already the case, the pass is enabled if those features are specified.

**Review by MaheshRavishankar — APPROVED** [review]
Thanks!

**Comment by dcaballe** [comment]
Thanks for the contribution! It definitely helps better understand the proposal. The RFC is currently under discussion so we should wait a bit for that to settle before moving forward with this.

**Comment by c-rhodes** [comment]
> Thanks for the contribution! It definitely helps better understand the proposal. The RFC is currently under discussion so we should wait a bit for that to settle before moving forward with this.

Thanks! I agree, best to wait for further comments.

**Comment by c-rhodes** [comment]
I've landed 2 patches in LLVM adding support for this:
* https://reviews.llvm.org/D150932
* https://reviews.llvm.org/D150934

and I've updated the IREE patch locally to use the pass in LLVM. I'll have to wait until those patches are in iree-llvm-fork until I can update this, what's the process there? I noticed there's a branch  `llvm-bump-20230524` bumping to yesterday.

**Comment by MaheshRavishankar** [comment]
> I've landed 2 patches in LLVM adding support for this:
> 
> * [reviews.llvm.org/D150932](https://reviews.llvm.org/D150932)
> * [reviews.llvm.org/D150934](https://reviews.llvm.org/D150934)
> 
> and I've updated the IREE patch locally to use the pass in LLVM. I'll have to wait until those patches are in iree-llvm-fork until I can update this, what's the process there? I noticed there's a branch `llvm-bump-20230524` bumping to yesterday.

Best option is to wait till the LLVM fork catches up. There is a process by which the LLVM commit IREE is using gets bumped, and there are a few moving pieces. You can follow along the https://discordapp.com/channels/689900678990135345/1080178290188374049 on discord to see whats happening.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUEnableAArch64SSVE.cpp [hunk 1]
```diff
@@ -0,0 +1,71 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+//===- LLVMCPUEnableAArch64SSVE.cpp ---------------------------------------===//
+//
+// This pass enables the Armv9 Scalable Matrix Extension (SME) Streaming SVE
+// (SSVE) mode [1].
```
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUEnableAArch64SSVE.cpp]: Should we move this to the pass class and use `///` so that it's picked it up by Doxygen?

**Review by dcaballe — APPROVED** [review]
LGTM, thanks!

**Comment by c-rhodes** [comment]
> > I've landed 2 patches in LLVM adding support for this:
> > 
> > * [reviews.llvm.org/D150932](https://reviews.llvm.org/D150932)
> > * [reviews.llvm.org/D150934](https://reviews.llvm.org/D150934)
> > 
> > and I've updated the IREE patch locally to use the pass in LLVM. I'll have to wait until those patches are in iree-llvm-fork until I can update this, what's the process there? I noticed there's a branch `llvm-bump-20230524` bumping to yesterday.
> 
> Best option is to wait till the LLVM fork catches up. There is a process by which the LLVM commit IREE is using gets bumped, and there are a few moving pieces. You can follow along the https://discordapp.com/channels/689900678990135345/1080178290188374049 on discord to see whats happening.

Ok cheers I'll update this PR once the fork catches up.

**Comment by c-rhodes** [comment]
Hi all, I've updated the patch to integrate the MLIR pass now the patches are in the fork, cheers.

**Comment by MaheshRavishankar** [comment]
> Hi all, I've updated the patch to integrate the MLIR pass now the patches are in the fork, cheers.

Thanks. Its already approved... I can land after CI passes.

**Comment by c-rhodes** [comment]
> > Hi all, I've updated the patch to integrate the MLIR pass now the patches are in the fork, cheers.
> 
> Thanks. Its already approved... I can land after CI passes.

Ok thanks. I've pushed a fix for the bazel_to_cmake CI failure.

**Comment by MaheshRavishankar** [comment]
> > > Hi all, I've updated the patch to integrate the MLIR pass now the patches are in the fork, cheers.
> > 
> > 
> > Thanks. Its already approved... I can land after CI passes.
> 
> Ok thanks. I've pushed a fix for the bazel_to_cmake CI failure.

@c-rhodes apologies for this being harder to land. Since this is your first commit, someone has to explicitly push a button to run the CI... After your first commit, you wont need that to run the CI (though someone will still need to land it for you, but still at least you can look at all CI failures independently).

**Comment by c-rhodes** [comment]
> > > > Hi all, I've updated the patch to integrate the MLIR pass now the patches are in the fork, cheers.
> > > 
> > > 
> > > Thanks. Its already approved... I can land after CI passes.
> > 
> > 
> > Ok thanks. I've pushed a fix for the bazel_to_cmake CI failure.
> 
> @c-rhodes apologies for this being harder to land. Since this is your first commit, someone has to explicitly push a button to run the CI... After your first commit, you wont need that to run the CI (though someone will still need to land it for you, but still at least you can look at all CI failures independently).

No worries, @MaheshRavishankar @dcaballe thank you both!

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/pipeline_tests.mlir [hunk 2]
```diff
@@ -602,3 +602,103 @@ hal.executable private @ukernel_pass_through {
...
+hal.executable private @aarch64_ssve__cpu_buffer_ops_tile_and_vectorize {
+  hal.executable.variant public @embedded_elf_arm_64, target = #executable_target_embedded_elf_arm_64_ {
+    hal.executable.export public @dispatch ordinal(0) layout(#pipeline_layout) attributes {
+      translation_info = #iree_codegen.translation_info<CPUBufferOpsTileAndVectorize>
+    } {
+    ^bb0(%arg0: !hal.device, %arg1: index, %arg2: index):
+      hal.return %arg1, %arg2, %arg2 : index, index, index
+    }
+    builtin.module {
+      func.func @dispatch() { return }
```
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/test/pipeline_tests.mlir:637]: Currently hitting a problem with these tests as they go through the different experts but there is no linalg op and, therefore, we don't generate any lowering config. A couple of suggestions:
  1. Could we add at least one linalg operation?
  2. Would it make sense to move the code that adds the SME pass to a common place for all the pipelines? Perhaps to the LowerToLLVM pass or at the end of LLVMCPULowerExecutableTargetPass::runOnOperation?
  ```
    if (enableAArch64SSVE)
    nestedModulePM.addNestedPass<func::FuncOp>(
        mlir::arm_sme::createEnableArmStreamingPass(
            mlir::arm_sme::ArmStreaming::Locally));
  ```
  to a common place for all the pipelines? Perhaps to the LowerToLLVM pass or at the end of LLVMCPULowerExecutableTargetPass::runOnOperation?

Comment (c-rhodes) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/pipeline_tests.mlir:637]: Hi Diego, I'm at a conference today but I'll try take a look later this afternoon.

Comment (c-rhodes) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/pipeline_tests.mlir:637]: > Currently hitting a problem with these tests as they go through the different experts but there is no linalg op and, therefore, we don't generate any lowering config. 

Could we not just add a lowering_config to the test like in the `@preset_pad_config_dynamic_matmul` test?

>Would it make sense to move the code that adds the SME pass to a common place for all the pipelines? Perhaps to the LowerToLLVM pass or at the end of LLVMCPULowerExecutableTargetPass::runOnOperation?

In the RFC I mentioned heuristics for enabling SSVE and how the existing pipeline structure in IREE could be leveraged to achieve this by selectively enabling SSVE for select ops that use select pipelines. I enabled SSVE for:
* CPUBufferOpsTileAndVectorize
* CPUDoubleTilingPeelingExpert
* CPUConvTileAndDecomposeExpert

to demonstrate this, perhaps a bit prematurely as I don't have concrete results that shows it's optimal for these pipelines, but we have a CV pipeline where all the ops use ones of these.

IIUC you're suggesting just enabling SSVE for all pipelines?

Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/pipeline_tests.mlir:637]: The problem is that the dispatches in these tests are empty and it looks like our pipelines haven't been taught to deal with that case. I guess it makes sense because if the dispatch is empty it shouldn't have been created. We can just add a dummy linalg op to the tests to fix the problem.

Regarding the location of the SSVE pass, you know better what else will be needed but, currently, it is just about introducing the attributes at function level after all the pipeline transformations has happen. That looks like generic to me and not so much pipeline specific so perhaps moving it to `addLowerToLLVMPasses` makes sense. We have plenty of post-pipeline transformations there. Since the SSVE pass is under a flag, it should be fine but, again, you know better :). This is just a suggestion. Nothing to worry about right now if we add a dummy linalg op to the tests.

Comment (banach-space) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/pipeline_tests.mlir:637]: Sorry that this is causing an issue. The content of these dispatches is irrelevant from the point of view what's being tested, so this makes a lot of sense:

> Could we add at least one linalg operation?

https://github.com/openxla/iree/pull/14099

> Regarding the location of the SSVE pass, you know better what else will be needed but, currently, it is just about introducing the attributes at function level after all the pipeline transformations has happen. That looks like generic to me and not so much pipeline specific 

The plan is to refine this, but we haven't started experimenting yet :( That's one of major tasks for us for the summer ™️ . We didn't want to enable this "everywhere" as there are certain instructions in the Streaming mode that become illegal (e.g. gather loads). And we do know that the backend still cannot handle some of the corner cases.

Tl;Dr We definitely need to revisit and to refine this. Thanks for flagging up Diego!
